src/chunkformatter.js

```diff
@@ -0,0 +1,311 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+const Mutation = require('./mutation.js');
+/**
+ * Enum for chunk formatter Row state.
+ * NEW_ROW: inital state or state after commitRow or resetRow
+ * ROW_IN_PROGRESS: state after first valid chunk without commitRow or resetRow
+ * CELL_IN_PROGRESS: state when valueSize > 0(partial cell)
+ */
+const RowStateEnum = Object.freeze({
+  NEW_ROW: 1,
+  ROW_IN_PROGRESS: 2,
+  CELL_IN_PROGRESS: 3,
+});
+/**
+ * ChunkFormatter formats all incoming chunks in to row
+ * keeps all intermediate state until end of stream.
+ * Should use new instance for each request.
+ */
+function ChunkFormatter() {
+  if (!(this instanceof ChunkFormatter)) {
+    return new ChunkFormatter();
+  }
+  this.reset();
+}
+/**
+ * Formats the row chunks into friendly format. Chunks contain 3 properties:
+ *
+ * `rowContents` The row contents, this essentially is all data pertaining
+ *     to a single family.
+ *
+ * `commitRow` This is a boolean telling us the all previous chunks for this
+ *     row are ok to consume.
+ *
+ * `resetRow` This is a boolean telling us that all the previous chunks are to
+ *     be discarded.
+ *
+ * @public
+ *
+ * @param {chunk[]} chunks The list of chunks.
+ * @param {object} [options] Formatting options.
+ * @param {callback} callback callback will be called with parsed rows
+ * @throws Error if it detects invalid state
+ * @example
+ * ChunkFormatter.formatChunks(chunks);
+ * // {
+ * //   follows: {
+ * //     gwashington: [
+ * //       {
+ * //         value: 2
+ * //       }
+ * //     ]
+ * //   }
+ * // }
+ */
+ChunkFormatter.prototype.formatChunks = function(chunks, options, callback) {
+  options = options || {};
+  for (const chunk of chunks) {
+    switch (this.state) {
+      case RowStateEnum.NEW_ROW:
+        this.newRow(chunk, options, callback);
+        break;
+      case RowStateEnum.ROW_IN_PROGRESS:
+        this.rowInProgress(chunk, options, callback);
+        break;
+      case RowStateEnum.CELL_IN_PROGRESS:
+        this.cellInProgress(chunk, options, callback);
+        break;
+    }
+  }
+};
+/**
+ * should be called at end of the stream to check if there is any pending row.
+ * @public
+ * @throws Error if there is any uncommitted row.
+ */
+ChunkFormatter.prototype.onStreamEnd = function() {
+  if (typeof this.row.key !== 'undefined') {
+    throw new Error('Response ended with pending row without commit');
+  }
+};
+/**
+ * Resets state of formatter
+ * @private
+ */
+ChunkFormatter.prototype.reset = function() {
+  this.prevRowKey = '';
```

Comment:
Shouldn't this be either undefined or null. It can be string or raw bytes depending on `decode` value.

Reply:
previous row key is assigned the row key of committed row which is always converted from bytes. Row key should never be blank even.

Reply:
Only reason, I was saying that if one look at [CellChunk::row-key] (https://github.com/googleapis/nodejs-bigtable/blob/master/protos/google/bigtable/v2/bigtable.proto#L105) datatype is bytes.